File: parseops.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Xml;
using System.Text;

namespace OpsParser
{
	public class OpsInfo
	{
		public OpsInfo()
		{
			asset = "";
			name = "";
			preferredasset1 = "";
			preferredasset2 = "";
			pos = new double[3];
			rot = new double[4];
			scl = new double[3];
			scl[0] = scl[1] = scl[2] = 1;
			materialdiffuse = new double[4];
			materialdiffuse[0] = materialdiffuse[1] = materialdiffuse[2] = materialdiffuse[3] = 1;
			materialemission = new double[3];
		}
		public String asset;
		public String name;
		public String preferredasset1;
		public String preferredasset2;
		// Order is x, y, z
		public double[] pos;
		public double[] rot; // in quaterion
		public double[] scl;
		public double[] materialdiffuse;
		public double[] materialemission;
	}

	public static class OpsParserProgram
	{
		static void parse_commadouble(String str, double[] arr)
		{
			var split = str.Replace(" ", "").Split(new char[] { ',' });
			for (var i = 0; i < split.Length; i += 1)
			{
				String split_no_period = split[i];
				if (split_no_period.Substring(split_no_period.Length - 1, 1) == ".")
				{
					split_no_period = split_no_period.Substring(0, split_no_period.Length - 1);
				}
				arr[i] = double.Parse(split_no_period, System.Globalization.CultureInfo.InvariantCulture);
			}
		}
		static void arr_rad_to_deg(double[] arr)
		{
			for (var i = 0; i < arr.Length; i += 1)
			{
				arr[i] = (arr[i] / (2 * Math.PI)) * 360;
			}
		}
		static void arr_deg_to_rad(double[] arr)
		{
			for (var i = 0; i < arr.Length; i += 1)
			{
				arr[i] = (arr[i] / 360) * (2 * Math.PI);
			}
		}
		static void arr_rad_to_quat(double[] arr)
		{
			var cy = Math.Cos(arr[2] * 0.5);
			var sy = Math.Sin(arr[2] * 0.5);
			var cp = Math.Cos(arr[1] * 0.5);
			var sp = Math.Sin(arr[1] * 0.5);
			var cr = Math.Cos(arr[0] * 0.5);
			var sr = Math.Sin(arr[0] * 0.5);
			arr[0] = cr * cp * cy + sr * sp * sy;
			arr[1] = sr * cp * cy - cr * sp * sy;
			arr[2] = cr * sp * cy + sr * cp * sy;
			arr[3] = cr * cp * sy - sr * sp * cy;
		}
		static void arr_normalize_deg(double[] arr)
		{
			for (var i = 0; i < arr.Length; i += 1)
			{
				arr[i] %= 360;
				if (arr[i] < 0)
				{
					arr[i] += 360;
				}
			}
		}
		static void arr_deg(double[] arr)
		{
			for (var i = 0; i < arr.Length; i += 1)
			{
				arr[i] = (arr[i] / (2 * Math.PI)) * 360;
			}
		}
		static void process_asset_xml(OpsInfo ops_info)
		{
			String fn = ops_info.asset;
			ops_info.preferredasset1 = fn;
			// special exceptions
			// TODO: add special exceptions for ED85
			if (fn == "O_C03TBL00")
			{
				ops_info.preferredasset2 = "c0tbl00";
				return;
			}
			if (fn == "O_C03TBL01")
			{
				ops_info.preferredasset2 = "c0tbl01";
				return;
			}
			if (fn == "O_C03TBL02")
			{
				ops_info.preferredasset2 = "c0tbl02";
				return;
			}
			if (fn == "O_C24DOR01")
			{
				ops_info.preferredasset2 = "24dor01";
				return;
			}
			if (fn == "O_F30KAG28")
			{
				ops_info.preferredasset2 = "s30kag28";
				return;
			}
			if (fn == "O_F44EVT00_GS")
			{
				ops_info.preferredasset2 = "f44evt_coaster_gs";
				return;
			}
			if (fn == "O_M21ETC01")
			{
				ops_info.preferredasset2 = "m21ect01";
				return;
			}
			if (fn == "O_M60EVT47")
			{
				ops_info.preferredasset2 = "s00evt47";
				return;
			}
			if (fn == "O_R02LIG00")
			{
				ops_info.preferredasset2 = "o_r02lig00";
				return;
			}
			if (fn == "O_S00FLS00")
			{
				ops_info.preferredasset2 = "light00";
				return;
			}
			if (fn == "O_S00SKYW")
			{
				ops_info.preferredasset2 = "s00skyw_sky";
				return;
			}
			if (fn == "O_S50KMO50")
			{
				ops_info.preferredasset2 = "s50obj50";
				return;
			}
			if (fn == "O_S61KMO45B")
			{
				ops_info.preferredasset2 = "s61kag45b";
				return;
			}
			if (fn == "O_S61KMO46")
			{
				ops_info.preferredasset2 = "s61kag46";
				return;
			}
			if (fn == "O_T00EVT61")
			{
				ops_info.preferredasset2 = "t00ev61";
				return;
			}
			if (fn == "O_T03EVT00")
			{
				ops_info.preferredasset2 = "t02evt00";
				return;
			}
			if (fn == "O_T10TIG03")
			{
				ops_info.preferredasset2 = "t10lig03";
				return;
			}
			if (fn == "O_T40KMO01")
			{
				ops_info.preferredasset2 = "t40lmo01";
				return;
			}
			if (fn == "O_T50KMO01H")
			{
				ops_info.preferredasset2 = "t50kag01h";
				return;
			}
			if (fn == "O_TESTRIGD")
			{
				ops_info.preferredasset2 = "rigidtest";
				return;
			}
			if (fn == "O_V03TRN02")
			{
				ops_info.preferredasset2 = "v03tm02";
				return;
			}
			// chop off O_
			fn = fn.Substring(2);
			// lowercase
			fn = fn.ToLower();
			ops_info.preferredasset2 = fn;
		}
		public static void read_xml(string filename, LinkedList<OpsInfo> ops_list)
		{
			var xml_doc = new XmlDocument();
			xml_doc.Load(filename);

			var assetObjects = xml_doc.GetElementsByTagName("AssetObject");

			for (var i = 0; i < assetObjects.Count; i += 1)
			{
				if (assetObjects[i] != null && assetObjects[i].Attributes != null)
				{
					var attr = assetObjects[i].Attributes;
					if (attr["asset"] != null && attr["name"] != null && attr["pos"] != null && attr["rot"] != null && attr["scl"] != null)
					{
						var opsInfo = new OpsInfo();
						opsInfo.asset = attr["asset"].InnerText;
						opsInfo.name = attr["name"].InnerText;
						parse_commadouble(attr["pos"].InnerText, opsInfo.pos);
						parse_commadouble(attr["rot"].InnerText, opsInfo.rot);
						parse_commadouble(attr["scl"].InnerText, opsInfo.scl);
						parse_commadouble(attr["materialDiffuse"].InnerText, opsInfo.materialdiffuse);
						parse_commadouble(attr["materialEmission"].InnerText, opsInfo.materialemission);
						// reverse X position
						opsInfo.pos[0] = -opsInfo.pos[0];
						// subtract 90 rotation (radian)
						opsInfo.rot[0] -= (Math.PI * 2) / 4;
						// reverse Y rotation (radian)
						opsInfo.rot[1] = -opsInfo.rot[1];
						arr_rad_to_quat(opsInfo.rot);
						process_asset_xml(opsInfo);
						ops_list.AddLast(opsInfo);
					}
					
				}
			}
		}
		public static void read_plt(string filename, LinkedList<OpsInfo> ops_list)
		{
			using (var reader = new BinaryReader(File.Open(filename, FileMode.Open)))
			{
				// skip null bytes
				reader.BaseStream.Seek(4, SeekOrigin.Current);
				var count = reader.ReadUInt32();

				byte[] filename_entry = new byte[32];
				float[] mat = new float[16];
				for (var i = 0; i < count; i += 1)
				{
					var objname = "";
					reader.Read(filename_entry, 0, 32);
					for (var j = 0; j < 32; j += 1)
					{
						if (filename_entry[j] == 0)
						{
							objname = System.Text.Encoding.ASCII.GetString(filename_entry, 0, j);
							break;
						}
					}
					var assname = "";
					reader.Read(filename_entry, 0, 32);
					for (var j = 0; j < 32; j += 1)
					{
						if (filename_entry[j] == 0)
						{
							assname = System.Text.Encoding.ASCII.GetString(filename_entry, 0, j);
							break;
						}
					}
					var unk1 = reader.ReadUInt32();
					var unk2 = reader.ReadUInt32();
					var unk3 = reader.ReadUInt32();
					var transformcount = reader.ReadUInt32();
					for (var j = 0; j < transformcount; j += 1)
					{
						for (var k = 0; k < 16; k += 1)
						{
							mat[k] = reader.ReadSingle();
						}
						var opsInfo = new OpsInfo();
						opsInfo.asset = assname;
						opsInfo.name = objname;
						opsInfo.pos[0] = mat[12];
						opsInfo.pos[1] = mat[13];
						opsInfo.pos[2] = mat[14];
						opsInfo.scl[0] = Math.Sqrt(Math.Pow(mat[0], 2) + Math.Pow(mat[4], 2) + Math.Pow(mat[8], 2));
						opsInfo.scl[1] = Math.Sqrt(Math.Pow(mat[1], 2) + Math.Pow(mat[5], 2) + Math.Pow(mat[9], 2));
						opsInfo.scl[2] = Math.Sqrt(Math.Pow(mat[2], 2) + Math.Pow(mat[6], 2) + Math.Pow(mat[10], 2));
						var tr = mat[0] + mat[5] + mat[10];
						if (tr > 0)
						{
							var S = Math.Sqrt(tr + 1.0) * 2; // S=4*qw
							opsInfo.rot[0] = ( 0.25 * S );
							opsInfo.rot[1] = ( (mat[9] - mat[6]) / S );
							opsInfo.rot[2] = ( (mat[2] - mat[8]) / S );
							opsInfo.rot[3] = ( (mat[4] - mat[1]) / S );
						}
						else if ((mat[0] > mat[5]) && (mat[0] > mat[10]))
						{
							var S = Math.Sqrt(1.0 + mat[0] - mat[5] - mat[10]) * 2; // S=4*qx
							opsInfo.rot[0] = ( (mat[9] - mat[6]) / S );
							opsInfo.rot[1] = ( 0.25 * S );
							opsInfo.rot[2] = ( (mat[1] + mat[4]) / S );
							opsInfo.rot[3] = ( (mat[2] + mat[8]) / S );
						}
						else if (mat[5] > mat[10])
						{
							var S = Math.Sqrt(1.0 + mat[5] - mat[0] - mat[10]) * 2; // S=4*qy
							opsInfo.rot[0] = ( (mat[2] - mat[8]) / S );
							opsInfo.rot[1] = ( (mat[1] + mat[4]) / S );
							opsInfo.rot[2] = ( 0.25 * S );
							opsInfo.rot[3] = ( (mat[6] + mat[9]) / S );
						}
						else
						{
							var S = Math.Sqrt(1.0 + mat[10] - mat[0] - mat[5]) * 2; // S=4*qz
							opsInfo.rot[0] = ( (mat[4] - mat[1]) / S );
							opsInfo.rot[1] = ( (mat[2] + mat[8]) / S );
							opsInfo.rot[2] = ( (mat[6] + mat[9]) / S );
							opsInfo.rot[3] = ( 0.25 * S );
						}
						// reverse X position
						opsInfo.pos[0] = -opsInfo.pos[0];
						// reverse X rotation (quat)
						opsInfo.rot[1] = -opsInfo.rot[1];
						// reverse Y rotation (quat)
						//opsInfo.rot[2] = -opsInfo.rot[2];
						process_asset_xml(opsInfo);
						ops_list.AddLast(opsInfo);
					}
				}
			}
		}
		static void process_asset_ed6(OpsInfo ops_info)
		{
			String fn = ops_info.asset;
			// chop off .x
			if (fn.Length >= 2)
			{
				fn = fn.Substring(0, fn.Length - 2);
			}
			ops_info.preferredasset1 = fn;
			// lowercase
			fn = fn.ToLower();
			ops_info.preferredasset2 = fn;
		}
		public static void read_ed6(string filename, LinkedList<OpsInfo> ops_list, bool is_ed6_3)
		{
			var opslength = 312;
			if (is_ed6_3)
			{
				opslength = 332;
			}
			using (var reader = new BinaryReader(File.Open(filename, FileMode.Open)))
			{
				if (is_ed6_3)
				{
					// skip header
					reader.BaseStream.Seek(20, SeekOrigin.Current);
				}
				var count = reader.ReadUInt16();
				byte[] filename_entry = new byte[12];
				byte[] entry = new byte[opslength - 12];
				float[] entry_floats = new float[entry.Length / sizeof(float)];
				for (var i = 0; i < count; i += 1)
				{
					var mdl_filename = "";
					reader.Read(filename_entry, 0, 12);
					for (var j = 0; j < 12; j += 1)
					{
						if (filename_entry[j] == 0)
						{
							mdl_filename = System.Text.Encoding.ASCII.GetString(filename_entry, 0, j);
							break;
						}
					}
					reader.Read(entry, 0, opslength - 12);
					for (var j = 0; j < entry_floats.Length; j += 1)
					{
						entry_floats[j] = BitConverter.ToSingle(entry, j * sizeof(float));
					}

					var opsInfo = new OpsInfo();
					opsInfo.asset = mdl_filename;
					opsInfo.pos[0] = entry_floats[63];
					opsInfo.pos[1] = entry_floats[62];
					opsInfo.pos[2] = entry_floats[61];
					opsInfo.rot[1] = entry_floats[64];
					arr_deg_to_rad(opsInfo.rot);
					arr_rad_to_quat(opsInfo.rot);
					process_asset_ed6(opsInfo);
					ops_list.AddLast(opsInfo);
				}
			}
		}
		static void Main(string[] args)
		{
			var path = args[0];

			LinkedList<OpsInfo> ops_list = new LinkedList<OpsInfo>();
			var is_xml = false;
			var is_plt = false;
			var is_ed6_3 = false;
			using (var reader = new BinaryReader(File.Open(path, FileMode.Open)))
			{
				var ident = reader.ReadUInt32();
				is_xml = ident == 0x3CBFBBEF;
				is_plt = ident == 0x00000000;
				is_ed6_3 = ident == 0x41364445;
			}
			
			if (is_xml)
			{
				read_xml(path, ops_list);
			}
			else if (is_plt)
			{
				read_plt(path, ops_list);
			}
			else
			{
				read_ed6(path, ops_list, is_ed6_3);
			}

			foreach (OpsInfo entry in ops_list)
			{
				Console.WriteLine("Asset: " + entry.asset);
				Console.WriteLine("Name: " + entry.name);
				Console.WriteLine("Prefname1: " + entry.preferredasset1);
				Console.WriteLine("Prefname2: " + entry.preferredasset2);
				Console.WriteLine("Position: " + entry.pos[0] + " " + entry.pos[1] + " " + entry.pos[2]);
				Console.WriteLine("Rotation: " + entry.rot[0] + " " + entry.rot[1] + " " + entry.rot[2] + " " + entry.rot[3]);
				Console.WriteLine("Scale: " + entry.scl[0] + " " + entry.scl[1] + " " + entry.scl[2]);
				Console.WriteLine("MaterialDiffuse: " + entry.materialdiffuse[0] + " " + entry.materialdiffuse[1] + " " + entry.materialdiffuse[2] + " " + entry.materialdiffuse[3]);
				Console.WriteLine("MaterialEmission: " + entry.materialemission[0] + " " + entry.materialemission[1] + " " + entry.materialemission[2]);
			}
		}
	}
}

File: parseops.py
import struct
import sys
import functools
import itertools
import array
import math
import os

import xml.etree.ElementTree as ET

global opsPrefabFolder
opsPrefabFolder = ""

class OpsInfo:
	def __init__(self):
		self.asset = "";
		self.name = "";
		self.preferredasset1 = "";
		self.preferredasset2 = "";
		# Order is x, y, z
		self.pos = [0.0, 0.0, 0.0]
		self.rot = [0.0, 0.0, 0.0, 0.0] # in quaterion
		self.scl = [1.0, 1.0, 1.0]
		self.materialdiffuse = [1.0, 1.0, 1.0, 1.0]
		self.materialemission = [0.0, 0.0, 0.0]

def readcstr(f):
    toeof = iter(functools.partial(f.read, 1), b'')
    return (b''.join(itertools.takewhile(b'\0'.__ne__, toeof))).decode("ASCII")

def parse_commadouble(s):
	lst = s.replace(" ", "").split(",")
	for i in range(len(lst)):
		lst[i] = float(lst[i])
	return lst

def arr_rad_to_deg(a):
	for i in range(len(a)):
		a[i] = (a[i] / (2 * math.pi)) * 360

def arr_deg_to_rad(a):
	for i in range(len(a)):
		a[i] = (a[i] / 360) * (2 * math.pi)

def arr_rad_to_quat(a):
	if len(a) < 4:
		a.append(0.0)
	cy = math.cos(a[2] * 0.5)
	sy = math.sin(a[2] * 0.5)
	cp = math.cos(a[1] * 0.5)
	sp = math.sin(a[1] * 0.5)
	cr = math.cos(a[0] * 0.5)
	sr = math.sin(a[0] * 0.5)
	a[0] = cr * cp * cy + sr * sp * sy
	a[1] = sr * cp * cy - cr * sp * sy
	a[2] = cr * sp * cy + sr * cp * sy
	a[3] = cr * cp * sy - sr * sp * cy

def arr_normalize_deg(a):
	for i in range(len(a)):
		a[i] %= 360
		if a[i] < 0:
			a[i] += 360

def arr_deg(a):
	for i in range(len(a)):
		a[i] = (a[i] / (2 * math.pi)) * 360

def arr_quat_to_mat3(m, q):
	sqrt2 = math.sqrt(2)
	q0 = sqrt2 * q[0]
	q1 = sqrt2 * q[1]
	q2 = sqrt2 * q[2]
	q3 = sqrt2 * q[3]

	qda = q0 * q1
	qdb = q0 * q2
	qdc = q0 * q3
	qaa = q1 * q1
	qab = q1 * q2
	qac = q1 * q3
	qbb = q2 * q2
	qbc = q2 * q3
	qcc = q3 * q3

	m[(0 * 3) + 0] = 1.0 - qbb - qcc
	m[(0 * 3) + 1] = qdc + qab
	m[(0 * 3) + 2] = -qdb + qac

	m[(1 * 3) + 0] = -qdc + qab
	m[(1 * 3) + 1] = 1.0 - qaa - qcc
	m[(1 * 3) + 2] = qda + qbc

	m[(2 * 3) + 0] = qdb + qac
	m[(2 * 3) + 1] = -qda + qbc
	m[(2 * 3) + 2] = 1.0 - qaa - qbb

def arr_mat3_normalized_to_eul2(mat, eul1, eul2):
	cy = math.hypot(mat[(0 * 3) + 0], mat[(0 * 3) + 1])

	epsilon = 1.192092896e-07
	if cy > 16.0 * epsilon:

		eul1[0] = math.atan2(mat[(1 * 3) + 2], mat[(2 * 3) + 2])
		eul1[1] = math.atan2(-mat[(0 * 3) + 2], cy)
		eul1[2] = math.atan2(mat[(0 * 3) + 1], mat[(0 * 3) + 0])

		eul2[0] = math.atan2(-mat[(1 * 3) + 2], -mat[(2 * 3) + 2])
		eul2[1] = math.atan2(-mat[(0 * 3) + 2], -cy)
		eul2[2] = math.atan2(-mat[(0 * 3) + 1], -mat[(0 * 3) + 0])
	else:
		eul1[0] = math.atan2(-mat[(2 * 3) + 1], mat[(1 * 3) + 1])
		eul1[1] = math.atan2(-mat[(0 * 3) + 2], cy)
		eul1[2] = 0.0

		eul2[0] = eul1[0]
		eul2[1] = eul1[1]
		eul2[2] = eul1[2]

def arr_mat3_normalized_to_eul(eul, mat):
	eul1 = [0.0, 0.0, 0.0]
	eul2 = [0.0, 0.0, 0.0]

	arr_mat3_normalized_to_eul2(mat, eul1, eul2)

	# return best, which is just the one with lowest values it in
	if (math.fabs(eul1[0]) + math.fabs(eul1[1]) + math.fabs(eul1[2])) > (math.fabs(eul2[0]) + math.fabs(eul2[1]) + math.fabs(eul2[2])):
		eul[0] = eul2[0]
		eul[1] = eul2[1]
		eul[2] = eul2[2]
	else:
		eul[0] = eul1[0]
		eul[1] = eul1[1]
		eul[2] = eul1[2]

def arr_quat_to_rad(a):
	unit_mat = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
	arr_quat_to_mat3(unit_mat, a)
	if len(a) == 4:
		del a[-1]
	arr_mat3_normalized_to_eul(a, unit_mat)

def process_asset_xml(ops_info):
	fn = ops_info.asset
	ops_info.preferredasset1 = fn
	# special exceptions
	if fn == "O_C03TBL00":
		ops_info.preferredasset2 = "c0tbl00"
		return
	if fn == "O_C03TBL01":
		ops_info.preferredasset2 = "c0tbl01"
		return
	if fn == "O_C03TBL02":
		ops_info.preferredasset2 = "c0tbl02"
		return
	if fn == "O_C24DOR01":
		ops_info.preferredasset2 = "24dor01"
		return
	if fn == "O_F30KAG28":
		ops_info.preferredasset2 = "s30kag28"
		return
	if fn == "O_F44EVT00_GS":
		ops_info.preferredasset2 = "f44evt_coaster_gs"
		return
	if fn == "O_M21ETC01":
		ops_info.preferredasset2 = "m21ect01"
		return
	if fn == "O_M60EVT47":
		ops_info.preferredasset2 = "s00evt47"
		return
	if fn == "O_R02LIG00":
		ops_info.preferredasset2 = "o_r02lig00"
		return
	if fn == "O_S00FLS00":
		ops_info.preferredasset2 = "light00"
		return
	if fn == "O_S00SKYW":
		ops_info.preferredasset2 = "s00skyw_sky"
		return
	if fn == "O_S50KMO50":
		ops_info.preferredasset2 = "s50obj50"
		return
	if fn == "O_S61KMO45B":
		ops_info.preferredasset2 = "s61kag45b"
		return
	if fn == "O_S61KMO46":
		ops_info.preferredasset2 = "s61kag46"
		return
	if fn == "O_T00EVT61":
		ops_info.preferredasset2 = "t00ev61"
		return
	if fn == "O_T03EVT00":
		ops_info.preferredasset2 = "t02evt00"
		return
	if fn == "O_T10TIG03":
		ops_info.preferredasset2 = "t10lig03"
		return
	if fn == "O_T40KMO01":
		ops_info.preferredasset2 = "t40lmo01"
		return
	if fn == "O_T50KMO01H":
		ops_info.preferredasset2 = "t50kag01h"
		return
	if fn == "O_TESTRIGD":
		ops_info.preferredasset2 = "rigidtest"
		return
	if fn == "O_V03TRN02":
		ops_info.preferredasset2 = "v03tm02"
		return
	# chop off O
	fn = fn[2:]
	# lowercase
	fn = fn.lower()
	ops_info.preferredasset2 = fn

def read_xml(f):
	lst = []
	root = ET.fromstring(f.read().decode("utf-8-sig"))
	for child in root:
		if child.tag == "MapObjects":
			for cchild in child:
				attr = cchild.attrib
				if "asset" in attr and "name" in attr and "pos" in attr and "rot" in attr and "scl" in attr:
					ops_info = OpsInfo()
					ops_info.asset = attr["asset"]
					ops_info.name = attr["name"]
					ops_info.pos = parse_commadouble(attr["pos"])
					ops_info.rot = parse_commadouble(attr["rot"])
					ops_info.scl = parse_commadouble(attr["scl"])
					ops_info.materialdiffuse = parse_commadouble(attr["materialDiffuse"])
					ops_info.materialemission = parse_commadouble(attr["materialEmission"])
					# reverse X position
					ops_info.pos[0] = -ops_info.pos[0]
					# subtract 90 rotation (radian)
					ops_info.rot[0] -= (math.pi * 2) / 4
					# reverse Y rotation (radian)
					ops_info.rot[1] = -ops_info.rot[1]
					arr_rad_to_quat(ops_info.rot)
					process_asset_xml(ops_info)
					lst.append(ops_info)
	return lst

def read_plt(f):
	lst = []
	shouldbezero = struct.unpack("I", f.read(4))
	if shouldbezero[0] != 0:
		raise "Should be zero"
	length = struct.unpack("I", f.read(4))
	if length[0] == 0:
		return
	for i in range(length[0]):
		oldpos = f.tell()
		objname = readcstr(f)
		f.seek(oldpos + 32)
		oldpos = f.tell()
		assname = readcstr(f)
		f.seek(oldpos + 32)
		unk1, unk2, unk3, transformlen = struct.unpack("IIII", f.read(16))
		print("Object: " + objname)
		print("Asset: " + assname)
		for j in range(transformlen):
			print("-----------------")
			arr = array.array("f", f.read(64))
			print(str(arr[0]) + " " + str(arr[1]) + " " + str(arr[2]) + " " + str(arr[3]))
			print(str(arr[4]) + " " + str(arr[5]) + " " + str(arr[6]) + " " + str(arr[7]))
			print(str(arr[8]) + " " + str(arr[9]) + " " + str(arr[10]) + " " + str(arr[11]))
			print(str(arr[12]) + " " + str(arr[13]) + " " + str(arr[14]) + " " + str(arr[15]))

def read_ed6(f):
	lst = []
	opslength = 312
	if f.read(20) == b"ED6AC_OPSFILE_VER2.0":
		opslength = 332
	else:
		f.seek(0)
	length = struct.unpack("H", f.read(2))
	if length[0] == 0:
		return
	for i in range(length[0]):
		oldpos = f.tell()
		filename = readcstr(f)
		f.seek(oldpos + 12)
		arr = array.array("f", f.read(opslength-12))
		print("Object: " + filename)
		print("X: " + str(arr[63]))
		print("Y: " + str(arr[62]))
		print("Z: " + str(arr[61]))
		print("Rotation: " + str(arr[64]))

for arg in sys.argv[1:]:
	with open(arg, "rb") as f:
		#print("FILE: " + arg)
		ident, = struct.unpack("I", f.read(4))
		f.seek(0)
		is_xml = ident == 0x3CBFBBEF
		is_plt = ident == 0x00000000
		is_ed6_3 = ident == 0x41364445
		lst = []
		if is_xml:
			lst = read_xml(f)
		elif is_plt:
			read_plt(f)
		else:
			read_ed6(f)
		if False:
			for entry in lst:
				print("Asset: " + entry.asset)
				print("Name: " + entry.name)
				print("Prefname1: " + entry.preferredasset1)
				print("Prefname2: " + entry.preferredasset2)
				print("Position: " + str(entry.pos[0]) + " " + str(entry.pos[1]) + " " + str(entry.pos[2]))
				print("Rotation: " + str(entry.rot[0]) + " " + str(entry.rot[1]) + " " + str(entry.rot[2]) + " " + str(entry.rot[3]))
				print("Scale: " + str(entry.scl[0]) + " " + str(entry.scl[1]) + " " + str(entry.scl[2]))
				print("MaterialDiffuse: " + str(entry.materialdiffuse[0]) + " " + str(entry.materialdiffuse[1]) + " " + str(entry.materialdiffuse[2]) + " " + str(entry.materialdiffuse[3]))
				print("MaterialEmission: " + str(entry.materialemission[0]) + " " + str(entry.materialemission[1]) + " " + str(entry.materialemission[2]))
		else:
			entries_deduped = {}
			for entry in lst:
				entries_deduped[entry.asset] = True
			with open(arg + ".sorted.txt", "w") as wf:
				for x in sorted(entries_deduped.keys()):
					if opsPrefabFolder == "" or (not os.path.isfile(opsPrefabFolder + "/" + x + ".prefab")):
						wf.write(x + "\n")

